File: src/dsl.rs
use crate::{
    ast::{query::Queriable, Circuit, StepType, StepTypeUUID},
    util::uuid,
    wit_gen::{FixedGenContext, StepInstance, TraceContext},
};

use halo2_proofs::plonk::{Advice, Column as Halo2Column, Fixed};

use core::fmt::Debug;
use std::marker::PhantomData;

use self::cb::{Constraint, LookupBuilder, Typing};

/// A generic structure designed to handle the context of a circuit for generic types `F`,
/// `TraceArgs` and `StepArgs`. The struct contains a `Circuit` instance and implements
/// methods to build the circuit, add various components, and manipulate the circuit. `F` is a
/// generic type representing the field of the circuit. `TraceArgs` is a generic type
/// representing the arguments passed to the trace function. `StepArgs` is a generic type
/// representing the arguments passed to the `step_type_def` function.
pub struct CircuitContext<F, TraceArgs> {
    sc: Circuit<F, TraceArgs>,
}

impl<F, TraceArgs> CircuitContext<F, TraceArgs> {
    /// Adds a forward signal to the circuit with a name string and zero rotation and returns a
    /// `Queriable` instance representing the added forward signal.
    pub fn forward(&mut self, name: &str) -> Queriable<F> {
        Queriable::Forward(self.sc.add_forward(name, 0), false)
    }

    /// Adds a forward signal to the circuit with a name string and a specified phase and returns a
    /// `Queriable` instance representing the added forward signal.
    pub fn forward_with_phase(&mut self, name: &str, phase: usize) -> Queriable<F> {
        Queriable::Forward(self.sc.add_forward(name, phase), false)
    }

    /// Adds a shared signal to the circuit with a name string and zero rotation and returns a
    /// `Queriable` instance representing the added shared signal.
    pub fn shared(&mut self, name: &str) -> Queriable<F> {
        Queriable::Shared(self.sc.add_shared(name, 0), 0)
    }

    /// Adds a shared signal to the circuit with a name string and a specified phase and returns a
    /// `Queriable` instance representing the added shared signal.
    pub fn shared_with_phase(&mut self, name: &str, phase: usize) -> Queriable<F> {
        Queriable::Shared(self.sc.add_shared(name, phase), 0)
    }

    pub fn fixed(&mut self, name: &str) -> Queriable<F> {
        Queriable::Fixed(self.sc.add_fixed(name), 0)
    }

    /// Exposes the first step instance value of a forward signal as public.
    pub fn expose(&mut self, forward: Queriable<F>) {
        match forward {
            Queriable::Forward(forward_signal, _) => {
                self.sc.expose(forward_signal);
            }
            _ => panic!("Expected forward signal"),
        }
    }

    /// Imports a halo2 advice column with a name string into the circuit and returns a
    /// `Queriable` instance representing the imported column.
    pub fn import_halo2_advice(&mut self, name: &str, column: Halo2Column<Advice>) -> Queriable<F> {
        Queriable::Halo2AdviceQuery(self.sc.add_halo2_advice(name, column), 0)
    }

    /// Imports a halo2 fixed column with a name string into the circuit and returns a
    /// `Queriable` instance representing the imported column.
    pub fn import_halo2_fixed(&mut self, name: &str, column: Halo2Column<Fixed>) -> Queriable<F> {
        Queriable::Halo2FixedQuery(self.sc.add_halo2_fixed(name, column), 0)
    }

    /// Adds a new step type with the specified name to the circuit and returns a
    /// `StepTypeHandler` instance. The `StepTypeHandler` instance can be used to define the
    /// step type using the `step_type_def` function.
    pub fn step_type(&mut self, name: &str) -> StepTypeHandler {
        let handler = StepTypeHandler::new(name.to_string());

        self.sc.add_step_type(handler, name);

        handler
    }

    /// Defines a step type using the provided `StepTypeHandler` and a function that takes a
    /// mutable reference to a `StepTypeContext`. This function typically adds constraints to a
    /// step type and defines witness generation.
    pub fn step_type_def<D, Args, S: Into<StepTypeDefInput>, R>(
        &mut self,
        step: S,
        def: D,
    ) -> StepTypeWGHandler<F, Args, R>
    where
        D: FnOnce(&mut StepTypeContext<F>) -> StepTypeWGHandler<F, Args, R>,
        R: Fn(&mut StepInstance<F>, Args) + 'static,
    {
        let handler: StepTypeHandler = match step.into() {
            StepTypeDefInput::Handler(h) => h,
            StepTypeDefInput::String(name) => {
                let handler = StepTypeHandler::new(name.to_string());

                self.sc.add_step_type(handler, name);

                handler
            }
        };

        let mut context = StepTypeContext::<F>::new(handler.uuid(), handler.annotation.to_string());

        let result = def(&mut context);

        self.sc.add_step_type_def(context.step_type);

        result
    }

    /// Sets the trace function that builds the witness. The trace function is responsible for
    /// adding step instances defined in `step_type_def`. The function is entirely left for
    /// the user to implement and is Turing complete. Users typically use external parameters
    /// of type `TraceArgs` to generate cell values for witness generation, and call the
    /// `add` function to add step instances with witness values.
    pub fn trace<D>(&mut self, def: D)
    where
        D: Fn(&mut TraceContext<F>, TraceArgs) + 'static,
    {
        self.sc.set_trace(def);
    }

    /// Sets the fixed generation function for the circuit. The fixed generation function is
    /// responsible for assigning fixed values to fixed columns. It is entirely left
    /// for the user to implement and is Turing complete. Users typically generate cell values and
    /// call the `assign` function to fill the fixed columns.
    pub fn fixed_gen<D>(&mut self, def: D)
    where
        D: Fn(&mut FixedGenContext<F>) + 'static,
    {
        self.sc.set_fixed_gen(def);
    }

    /// Enforce the type of the first step by adding a constraint to the circuit. Takes a
    /// `StepTypeHandler` parameter that represents the step type.
    pub fn pragma_first_step<STH: Into<StepTypeHandler>>(&mut self, step_type: STH) {
        self.sc.first_step = Some(step_type.into().uuid());
    }

    /// Enforce the type of the last step by adding a constraint to the circuit. Takes a
    /// `StepTypeHandler` parameter that represents the step type.
    pub fn pragma_last_step<STH: Into<StepTypeHandler>>(&mut self, step_type: STH) {
        self.sc.last_step = Some(step_type.into().uuid());
    }

    pub fn pragma_num_steps(&mut self, num_steps: usize) {
        self.sc.num_steps = num_steps;
    }
}

pub enum StepTypeDefInput {
    Handler(StepTypeHandler),
    String(&'static str),
}

impl From<StepTypeHandler> for StepTypeDefInput {
    fn from(h: StepTypeHandler) -> Self {
        StepTypeDefInput::Handler(h)
    }
}

impl From<&'static str> for StepTypeDefInput {
    fn from(s: &'static str) -> Self {
        StepTypeDefInput::String(s)
    }
}

/// A generic structure designed to handle the context of a step type definition.  The struct
/// contains a `StepType` instance and implements methods to build the step type, add components,
/// and manipulate the step type. `F` is a generic type representing the field of the step type.
/// `Args` is the type of the step instance witness generation arguments.
pub struct StepTypeContext<F> {
    step_type: StepType<F>,
}

impl<F> StepTypeContext<F> {
    pub fn new(uuid: u32, name: String) -> Self {
        Self {
            step_type: StepType::new(uuid, name),
        }
    }

    /// Adds an internal signal to the step type with the given name and returns a `Queriable`
    /// instance representing the added signal.
    pub fn internal(&mut self, name: &str) -> Queriable<F> {
        Queriable::Internal(self.step_type.add_signal(name))
    }

    /// DEPRECATED
    pub fn constr<C: Into<Constraint<F>>>(&mut self, constraint: C) {
        println!("DEPRECATED constr: use setup for constraints in step types");

        let constraint = constraint.into();

        self.step_type
            .add_constr(constraint.annotation, constraint.expr);
    }

    /// DEPRECATED
    pub fn transition<C: Into<Constraint<F>>>(&mut self, constraint: C) {
        println!("DEPRECATED transition: use setup for constraints in step types");

        let constraint = constraint.into();

        self.step_type
            .add_transition(constraint.annotation, constraint.expr);
    }

    /// Define step constraints.
    pub fn setup<D>(&mut self, def: D)
    where
        D: Fn(&mut StepTypeSetupContext<F>),
    {
        let mut ctx = StepTypeSetupContext {
            step_type: &mut self.step_type,
        };

        def(&mut ctx);
    }

    pub fn wg<Args, D>(&mut self, def: D) -> StepTypeWGHandler<F, Args, D>
    where
        D: Fn(&mut StepInstance<F>, Args) + 'static,
    {
        StepTypeWGHandler {
            id: self.step_type.uuid(),
            annotation: Box::leak(self.step_type.name.clone().into_boxed_str()),

            wg: Box::new(def),

            _p: PhantomData::default(),
        }
    }
}

impl<F: Debug + Clone> StepTypeContext<F> {
    /// DEPRECATED
    pub fn add_lookup(&mut self, lookup_builder: &mut LookupBuilder<F>) {
        println!("DEPRECATED add_lookup: use setup for constraints in step types");

        self.step_type.lookups.push(lookup_builder.lookup.clone());
    }
}

pub struct StepTypeSetupContext<'a, F> {
    step_type: &'a mut StepType<F>,
}

impl<'a, F> StepTypeSetupContext<'a, F> {
    /// Adds a constraint to the step type. Involves internal signal(s) and forward signals without
    /// SuperRotation only. Chiquito provides syntax sugar for defining complex constraints.
    /// Refer to the `cb` (constraint builder) module for more information.
    pub fn constr<C: Into<Constraint<F>>>(&mut self, constraint: C) {
        let constraint = constraint.into();
        Self::enforce_constraint_typing(&constraint);

        self.step_type
            .add_constr(constraint.annotation, constraint.expr);
    }

    /// Adds a transition constraint to the step type. It’s the same as a regular constraint except
    /// that it can involve forward signal(s) with SuperRotation as well. Chiquito provides syntax
    /// sugar for defining complex constraints. Refer to the `cb` (constraint builder) module
    /// for more information.
    pub fn transition<C: Into<Constraint<F>>>(&mut self, constraint: C) {
        let constraint = constraint.into();
        Self::enforce_constraint_typing(&constraint);

        self.step_type
            .add_transition(constraint.annotation, constraint.expr);
    }

    fn enforce_constraint_typing(constraint: &Constraint<F>) {
        if constraint.typing != Typing::AntiBooly {
            panic!(
                "Expected AntiBooly constraint, got {:?} (constraint: {})",
                constraint.typing, constraint.annotation
            );
        }
    }
}

impl<'a, F: Debug + Clone> StepTypeSetupContext<'a, F> {
    /// Adds a lookup to the step type.
    pub fn add_lookup(&mut self, lookup_builder: &mut LookupBuilder<F>) {
        self.step_type.lookups.push(lookup_builder.lookup.clone());
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub struct StepTypeHandler {
    id: StepTypeUUID,
    pub annotation: &'static str,
}

impl StepTypeHandler {
    fn new(annotation: String) -> Self {
        Self {
            id: uuid(),
            annotation: Box::leak(annotation.into_boxed_str()),
        }
    }

    pub fn uuid(&self) -> u32 {
        self.id
    }

    pub fn next<F>(&self) -> Queriable<F> {
        Queriable::StepTypeNext(*self)
    }
}

impl<F, Args, D: Fn(&mut StepInstance<F>, Args) + 'static> From<&StepTypeWGHandler<F, Args, D>>
    for StepTypeHandler
{
    fn from(h: &StepTypeWGHandler<F, Args, D>) -> Self {
        StepTypeHandler {
            id: h.id,
            annotation: h.annotation,
        }
    }
}

pub struct StepTypeWGHandler<F, Args, D: Fn(&mut StepInstance<F>, Args) + 'static> {
    id: u32,
    pub annotation: &'static str,
    pub wg: Box<D>,

    _p: PhantomData<(F, Args)>,
}

impl<F, Args, D: Fn(&mut StepInstance<F>, Args) + 'static> StepTypeWGHandler<F, Args, D> {
    pub fn uuid(&self) -> u32 {
        self.id
    }
}

/// Creates a `Circuit` instance by providing a name and a definition closure that is applied to a
/// mutable `CircuitContext`. The user customizes the definition closure by calling `CircuitContext`
/// functions. This is the main function that users call to define a Chiquito circuit. Currently,
/// the name is not used for annotation within the function, but it may be used in future
/// implementations.
pub fn circuit<F, TraceArgs, StepArgs, D>(_name: &str, def: D) -> Circuit<F, TraceArgs>
where
    D: Fn(&mut CircuitContext<F, TraceArgs>),
{
    // TODO annotate circuit
    let mut context = CircuitContext {
        sc: Circuit::default(),
    };

    def(&mut context);

    context.sc
}

pub mod cb;

File: src/frontend/python/chiquito/dsl.py
from __future__ import annotations
from typing import List, Dict
from enum import Enum
from typing import Callable, Any

from chiquito import rust_chiquito
import json

from chiquito.chiquito_ast import ASTCircuit, ASTStepType, ExposeOffset, ASTSuperCircuit
from chiquito.query import Internal, Forward, Queriable, Shared, Fixed
from chiquito.wit_gen import StepInstance, TraceWitness
from chiquito.cb import (
    Constraint,
    Typing,
    ToConstraint,
    to_constraint,
    LookupTable,
    LookupTableBuilder,
    InPlaceLookupBuilder,
)
from chiquito.util import CustomEncoder, F


class SuperCircuitMode(Enum):
    NoMode = 0
    SETUP = 1
    Mapping = 2


class SuperCircuit:
    def __init__(self: SuperCircuit):
        self.ast = ASTSuperCircuit()
        self.mode = SuperCircuitMode.SETUP
        self.setup()
        self.mode = SuperCircuitMode.NoMode

    # called under setup()
    def sub_circuit(self: SuperCircuit, sub_circuit: Circuit) -> Circuit:
        assert self.mode == SuperCircuitMode.SETUP
        if sub_circuit.rust_id != 0:
            raise ValueError(
                "SuperCircuit: sub_circuit() cannot be called twice on the same circuit."
            )
        ast_json: str = sub_circuit.get_ast_json()
        sub_circuit.rust_id: int = rust_chiquito.ast_to_halo2(ast_json)
        self.ast.sub_circuits[sub_circuit.rust_id] = sub_circuit.ast
        return sub_circuit

    # called under mapping()
    # generates TraceWitness for sub_circuit
    def map(self: SuperCircuit, sub_circuit: Circuit, *args: Any) -> TraceWitness:
        assert self.mode == SuperCircuitMode.Mapping
        witness: TraceWitness = sub_circuit.gen_witness(*args)
        if sub_circuit.rust_id == 0:
            raise ValueError(
                "SuperCircuit: must call sub_circuit() before calling map() on a Circuit."
            )
        self.ast.super_witness[sub_circuit.rust_id] = witness
        return witness

    # called at the outermost level
    # generates TraceWitness mapping
    def gen_witness(self: SuperCircuit, *args: Any) -> Dict[int, TraceWitness]:
        self.mode = SuperCircuitMode.Mapping
        self.mapping(*args)
        self.mode = SuperCircuitMode.NoMode
        super_witness: Dict[int, TraceWitness] = self.ast.super_witness
        del (
            self.ast.super_witness
        )  # so that we can generate different witness mapping in the next gen_witness() call
        return super_witness

    def halo2_mock_prover(self: SuperCircuit, super_witness: Dict[int, TraceWitness]):
        for rust_id, witness in super_witness.items():
            if rust_id not in self.ast.sub_circuits:
                raise ValueError(
                    f"SuperCircuit.halo2_mock_prover(): TraceWitness with rust_id {rust_id} not found in sub_circuits."
                )
            witness_json: str = witness.get_witness_json()
            super_witness[rust_id] = witness_json
        rust_chiquito.super_circuit_halo2_mock_prover(
            list(self.ast.sub_circuits.keys()), super_witness
        )


class CircuitMode(Enum):
    NoMode = 0
    SETUP = 1
    Trace = 2
    FixedGen = 3


class Circuit:
    def __init__(
        self: Circuit,
        super_circuit: SuperCircuit = None,
        **kwargs,  # **kwargs is intended for arbitrary names for imports
    ):
        self.ast = ASTCircuit()
        self.witness = TraceWitness()
        self.rust_id = 0
        self.super_circuit = super_circuit
        for key, value in kwargs.items():
            setattr(self, key, value)
        self.mode = CircuitMode.SETUP
        self.setup()
        if hasattr(self, "fixed_gen") and callable(self.fixed_gen):
            self.mode = CircuitMode.FixedGen
            if self.ast.num_steps == 0:
                raise ValueError(
                    "Must set num_steps by calling pragma_num_steps() in setup before calling fixed_gen()."
                )
            self.ast.fixed_assignments = {}
            self.fixed_gen()
        self.mode = CircuitMode.NoMode

    def forward(self: Circuit, name: str) -> Forward:
        assert self.mode == CircuitMode.SETUP
        return Forward(self.ast.add_forward(name, 0), False)

    def forward_with_phase(self: Circuit, name: str, phase: int) -> Forward:
        assert self.mode == CircuitMode.SETUP
        return Forward(self.ast.add_forward(name, phase), False)

    def shared(self: Circuit, name: str) -> Shared:
        assert self.mode == CircuitMode.SETUP
        return Shared(self.ast.add_shared(name, 0), 0)

    def shared_with_phase(self: Circuit, name: str, phase: int) -> Shared:
        assert self.mode == CircuitMode.SETUP
        return Shared(self.ast.add_shared(name, phase), 0)

    def fixed(self: Circuit, name: str) -> Fixed:
        assert self.mode == CircuitMode.SETUP
        return Fixed(self.ast.add_fixed(name), 0)

    def expose(self: Circuit, signal: Queriable, offset: ExposeOffset):
        assert self.mode == CircuitMode.SETUP
        if isinstance(signal, (Forward, Shared)):
            self.ast.expose(signal, offset)
        else:
            raise TypeError(f"Can only expose ForwardSignal or SharedSignal.")

    def step_type(self: Circuit, step_type: StepType) -> StepType:
        assert self.mode == CircuitMode.SETUP
        self.ast.add_step_type(step_type.step_type, step_type.step_type.name)
        return step_type

    def step_type_def(self: StepType) -> StepType:
        assert self.mode == CircuitMode.SETUP
        self.ast.add_step_type_def()

    def pragma_first_step(self: Circuit, step_type: StepType) -> None:
        assert self.mode == CircuitMode.SETUP
        self.ast.first_step = step_type.step_type.id

    def pragma_last_step(self: Circuit, step_type: StepType) -> None:
        assert self.mode == CircuitMode.SETUP
        self.ast.last_step = step_type.step_type.id

    def pragma_num_steps(self: Circuit, num_steps: int) -> None:
        assert self.mode == CircuitMode.SETUP
        self.ast.num_steps = num_steps

    def pragma_disable_q_enable(self: Circuit) -> None:
        assert self.mode == CircuitMode.SETUP
        self.ast.q_enable = False

    def new_table(self: Circuit, table: LookupTable) -> LookupTable:
        assert self.mode == CircuitMode.SETUP
        # have a method called set_finished_flag() to encapsulate
        # call finished_flag "finished" instead
        table.set_finished_flag()
        return table

    # called under trace()
    def add(self: Circuit, step_type: StepType, *args: Any):
        assert self.mode == CircuitMode.Trace
        if len(self.witness.step_instances) >= self.ast.num_steps:
            raise ValueError(f"Number of step instances exceeds {self.ast.num_steps}")
        step_instance: StepInstance = step_type.gen_step_instance(*args)
        self.witness.step_instances.append(step_instance)

    def needs_padding(self: Circuit) -> bool:
        assert self.mode == CircuitMode.Trace
        return len(self.witness.step_instances) < self.ast.num_steps

    def padding(self: Circuit, step_type: StepType, *args):
        assert self.mode == CircuitMode.Trace
        while self.needs_padding():
            self.add(step_type, *args)

    # called under fixed_gen()
    def assign(self: Circuit, offset: int, lhs: Queriable, rhs: F):
        assert self.mode == CircuitMode.FixedGen
        if self.ast.fixed_assignments is None:
            raise ValueError(
                "FixedAssignment: must have initiated fixed_assignments before calling assign()"
            )
        self.ast.add_fixed_assignment(offset, lhs, F(rhs))

    def gen_witness(self: Circuit, *args) -> TraceWitness:
        self.mode = CircuitMode.Trace
        self.witness = TraceWitness()
        self.trace(*args)
        self.mode = CircuitMode.NoMode
        witness = self.witness
        del self.witness
        return witness

    def get_ast_json(self: Circuit) -> str:
        return json.dumps(self.ast, cls=CustomEncoder, indent=4)

    def halo2_mock_prover(self: Circuit, witness: TraceWitness):
        if self.rust_id == 0:
            ast_json: str = self.get_ast_json()
            self.rust_id: int = rust_chiquito.ast_to_halo2(ast_json)
        witness_json: str = witness.get_witness_json()
        rust_chiquito.halo2_mock_prover(witness_json, self.rust_id)

    def __str__(self: Circuit) -> str:
        return self.ast.__str__()


class StepTypeMode(Enum):
    NoMode = 0
    SETUP = 1
    WG = 2


class StepType:
    def __init__(self: StepType, circuit: Circuit, step_type_name: str):
        self.step_type = ASTStepType.new(step_type_name)
        self.circuit = circuit
        self.mode = StepTypeMode.SETUP
        self.setup()

    def gen_step_instance(self: StepType, *args) -> StepInstance:
        self.mode = StepTypeMode.WG
        self.step_instance = StepInstance.new(self.step_type.id)
        self.wg(*args)
        self.mode = StepTypeMode.NoMode
        step_instance = self.step_instance
        del self.step_instance
        return step_instance

    def internal(self: StepType, name: str) -> Internal:
        assert self.mode == StepTypeMode.SETUP

        return Internal(self.step_type.add_signal(name))

    def constr(self: StepType, constraint: ToConstraint):
        assert self.mode == StepTypeMode.SETUP

        constraint = to_constraint(constraint)
        StepType.enforce_constraint_typing(constraint)
        self.step_type.add_constr(constraint.annotation, constraint.expr)

    def transition(self: StepType, constraint: ToConstraint):
        assert self.mode == StepTypeMode.SETUP

        constraint = to_constraint(constraint)
        StepType.enforce_constraint_typing(constraint)
        self.step_type.add_transition(constraint.annotation, constraint.expr)

    def enforce_constraint_typing(constraint: Constraint):
        if constraint.typing != Typing.AntiBooly:
            raise ValueError(
                f"Expected AntiBooly constraint, got {constraint.typing} (constraint: {constraint.annotation})"
            )

    def assign(self: StepType, lhs: Queriable, rhs: F):
        assert self.mode == StepTypeMode.WG

        self.step_instance.assign(lhs, F(rhs))

    def add_lookup(self: StepType, lookup_builder: LookupBuilder):
        self.step_type.lookups.append(lookup_builder.build())


LookupBuilder = LookupTableBuilder | InPlaceLookupBuilder
